File: tree/tree.go
package main

import "fmt"

type (
	Node struct {
		Value *int
		Left *Node
		Right *Node
	}

	Tree struct {
		Root *Node
	}
)

func NewNode(value int) *Node {
	return &Node{ Value: &value }
}

func New(value int) *Tree {
	return &Tree{
		Root: &Node{
			Value: &value,
		},
	}
}

func deque(queue *[]*Node) *Node {
	queueVal := *queue
	first := queueVal[0]
	queueVal = queueVal[1:]
	*queue = queueVal
	return first
}

func enque(queue *[]*Node, val *Node) {
	queueVal := *queue
	queueVal = append(queueVal, val)
	*queue = queueVal
}

func (t *Tree) levelOrderTraversal() []int {
	queue := make([]*Node, 0)
	data := make([]int, 0)
	enque(&queue, t.Root)
	for (len(queue) > 0) {
		currentNode := deque(&queue)
		if currentNode != nil {
			data = append(data, *currentNode.Value)
			enque(&queue, currentNode.Left)
			enque(&queue, currentNode.Right)
		}
	}

	return data
}

// root -> left -> right
func (t *Tree) preOrderTraversal() []int {
	data := make([]int, 0)
	data = *t.preOrderTraversalHelper(t.Root, &data)
	return data
}

func (t *Tree) preOrderTraversalHelper(node *Node, data *[]int) *[]int {
	if (node == nil) {
		return data
	}
	
	*data = append(*data, *node.Value)
	t.preOrderTraversalHelper(node.Left, data)
	t.preOrderTraversalHelper(node.Right, data)
	return data
}

// left -> root -> right
func (t *Tree) inOrderTraversal() []int {
	data := make([]int, 0)
	data = *t.inOrderTraversalHelper(t.Root, &data)
	return data
}

func (t *Tree) inOrderTraversalHelper(node *Node, data *[]int) *[]int {
	if (node == nil) {
		return data
	}
	
	t.inOrderTraversalHelper(node.Left, data)
	*data = append(*data, *node.Value)
	t.inOrderTraversalHelper(node.Right, data)
	return data
}

// left -> right -> root
func (t *Tree) postOrderTraversal() []int {
	data := make([]int, 0)
	data = *t.postOrderTraversalHelper(t.Root, &data)
	return data
}

func (t *Tree) postOrderTraversalHelper(node *Node, data *[]int) *[]int {
	if (node == nil) {
		return data
	}
	
	t.postOrderTraversalHelper(node.Left, data)
	t.postOrderTraversalHelper(node.Right, data)
	*data = append(*data, *node.Value)
	return data
}

//          1
//        /   \
//       2     3
//     /  \   /  \
//    4    5 6    7
func main () {
	var tree Tree = *New(1)
	var leftNode = NewNode(2)
	leftNode.Left = NewNode(4)
	leftNode.Right = NewNode(5)
	var rightNode = NewNode(3)
	rightNode.Left = NewNode(6)
	rightNode.Right = NewNode(7)
	tree.Root.Left = leftNode
	tree.Root.Right = rightNode

	fmt.Println(tree.inOrderTraversal())
	fmt.Println(tree.preOrderTraversal())
	fmt.Println(tree.postOrderTraversal())
	fmt.Println(tree.levelOrderTraversal())
}

File: tree/tree.py
class Node:
    value = None
    left = None
    right = None

    def __init__(self, value, left = None, right = None):
        self.value = value
        self.right = right
        self.left = left

class Tree:
    root = None

    def __init__(self, value):
        self.root = Node(value)
    
    def level_order_traversal(self):
        queue = [self.root]
        data = []
        while (len(queue) > 0):
            current_node = queue.pop(0) # deque the queue
            if current_node:
                data.append(current_node.value)
                if current_node.left:
                    queue.append(current_node.left) # enque the left node
                if current_node.right:
                    queue.append(current_node.right) # enque the right node

        return data

    def __pre_order_traversal_helper(self, node, data = []):
        if node == None:
            return data
    
        data.append(node.value)
        self.__pre_order_traversal_helper(node.left, data)
        self.__pre_order_traversal_helper(node.right, data)

        return data
    
    def __in_order_traversal_helper(self, node, data = []):
        if node == None:
            return data
    
        self.__in_order_traversal_helper(node.left, data)
        data.append(node.value)
        self.__in_order_traversal_helper(node.right, data)

        return data
    
    def __post_order_traversal_helper(self, node, data = []):
        if node == None:
            return data
    
        self.__post_order_traversal_helper(node.left, data)
        self.__post_order_traversal_helper(node.right, data)
        data.append(node.value)

        return data
    
    def pre_order_traversal(self):
        return self.__pre_order_traversal_helper(self.root)
    
    def in_order_traversal(self):
        return self.__in_order_traversal_helper(self.root)
    
    def post_order_traversal(self):
        return self.__post_order_traversal_helper(self.root)
    
tree = Tree(1)
leftNode = Node(2)
leftNode.left = Node(4)
leftNode.right = Node(5)
rightNode = Node(3)
rightNode.left = Node(6)
rightNode.right = Node(7)
tree.root.left = leftNode
tree.root.right = rightNode

#         1
#       /   \
#      2     3
#    /  \   /  \
#   4    5 6    7

print(tree.in_order_traversal())
print(tree.pre_order_traversal())
print(tree.post_order_traversal())
print(tree.level_order_traversal())
